api/rosetta/data_balance_integration_test.go

```diff
@@ -177,98 +201,542 @@ func TestGetBalance(t *testing.T) {
+			// verify that we have the same amount of balances as requested
+			if assert.Equal(t, len(balanceResponse.Balances), len(test.wantBalances)) {
 
-				// nothing more to do, response validation should only be done for '200 OK' responses
-				return
+				// verify returned balances - both the value and that the output matches the input spec
+				for i, wanted := range test.wantBalances {
+
+					balance := balanceResponse.Balances[i]
+
+					assert.Equal(t, test.request.Currencies[i].Symbol, balance.Currency.Symbol)
+					assert.Equal(t, test.request.Currencies[i].Decimals, balance.Currency.Decimals)
+					assert.Equal(t, wanted, balance.Value)
+				}
 			}
+		})
+	}
+}
 
-			assert.Equal(t, test.wantStatusCode, rec.Result().StatusCode)
+func TestBalanceErrors(t *testing.T) {
 
-			// unpack response
-			var balanceResponse rosetta.BalanceResponse
-			require.NoError(t, json.Unmarshal(rec.Body.Bytes(), &balanceResponse))
+	db := setupDB(t)
+	api := setupAPI(t, db)
 
-			// verify that the block data matches the input data
-			assert.Equal(t, test.request.BlockID.Index, balanceResponse.BlockID.Index)
-			assert.Equal(t, test.request.BlockID.Hash, balanceResponse.BlockID.Hash)
+	// defined valid balance request fields
+	var (
+		testAccount = identifier.Account{Address: "754aed9de6197641"}
 
-			// verify that we have at least one balance in the response
-			if assert.Len(t, balanceResponse.Balances, 1) {
+		testBlock = identifier.Block{
+			Index: 13,
+			Hash:  "af528bb047d6cd1400a326bb127d689607a096f5ccd81d8903dfebbac26afb23",
+		}
 
-				// verify the balance data - both the value and that the output matches the input spec
-				balance := balanceResponse.Balances[0]
-				assert.Equal(t, test.request.Currencies[0].Symbol, balance.Currency.Symbol)
-				assert.Equal(t, test.request.Currencies[0].Decimals, balance.Currency.Decimals)
-				assert.Equal(t, test.wantBalance, balance.Value)
-			}
+		validAddressSize = 2 * flow.AddressLength
+	)
+
+	const (
+		invalidAddress = "0000000000000000" // valid 16-digit hex value but not a valid account ID
+
+		trimmedBlockHash = "af528bb047d6cd1400a326bb127d689607a096f5ccd81d8903dfebbac26afb2" // block hash a character short
+
+		trimmedAddress    = "754aed9de619764"  // account ID a character short
+		invalidAddressHex = "zzzzzzzzzzzzzzzz" // invalid hex string
+
+		accFirstOccurrence = 13
+
+		// FIXME: when it's checked, see how to handle this error;
+		// it's cadence internal and might change, but also something we want to check
+		kludgyCadenceRuntimeError = "could not invoke script: script execution encountered error: [Error Code: 1101] cadence runtime error Execution failed:\nerror: panic: Could not borrow Balance reference to the Vault\n  --> d6b84b6f36db7d880d4ecc2a6a952094301873657e204e86d4cc9282c0df4b3d:11:11\n   |\n11 |         ?? panic(\"Could not borrow Balance reference to the Vault\")\n   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n"
```

Comment:
Used in the `account does not exist yet` test case - we try to retrieve an account balance at a height before the account is created.

We fail correctly, but the error description returned is the cadence internal error.

How should we handle this nicely? For all of the test cases we use string matching for the error. I was thinking we could perhaps use a regex instead and still do full match on the others, but for this only check the start of the string, to e.g. match `could not invoke script: script execution encountered error` or something like that.

Or, I just leave the field empty and in the code do
```go
if test.wantRosettaErrorDescription != "" {
  assert.Equal(t, test.wantRosettaErrorDescription, err.Description)
}
```

But I guess the bigger question is - should the API relay this error at all?

Reply:
This is a good point; at the moment, we are unfortunately unable to differentiate between an account not existing, or an account not having a Flow vault. I do think we should string-match this and return a `404`, but document that it stands for the vault not being found, not the account.